### PR TITLE
Serialize (and deserialize) int64 values

### PR DESCRIPTION
**Pull request by rsgowman**

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 1]
```diff
@@ -94,6 +94,15 @@ bool DecodeBool(pb_istream_t* stream) {
   }
 }
 
+void EncodeLong(pb_ostream_t* stream, int64_t integer_value) {
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: Sorry I haven't noticed this before but three different names for this thing in a single line finally caught my eye :-).

These aren't "longs" and that name is misleading on 32-bit platforms (e.g. iPhone 5C which we still support) and on win64.

I see two paths forward:
  * Call them "Integer" like the Firestore proto does (you can see this in `google_firestore_v1beta1_Value_integer_value_tag`)
  * Call them Int64 like the actual type of the value

I slightly prefer the former.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 2]
```diff
@@ -152,6 +168,8 @@ FieldValue Serializer::DecodeFieldValue(
       return FieldValue::NullValue();
     case FieldValue::Type::Boolean:
       return FieldValue::BooleanValue(value_proto.value.boolean_value);
+    case FieldValue::Type::Long:
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: The mismatch is obvious here too: could you rename `FieldValue::Type::Long` to match above?

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc [hunk 3]
```diff
@@ -134,6 +136,76 @@ TEST_F(SerializerTest, EncodesBoolProtoToBytes) {
   }
 }
 
+TEST_F(SerializerTest, EncodesIntegersModelToProto) {
+  std::list<int64_t> testCases{0,
```
Comment (wilhuff) [inline Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: I would imagine our default collection would just be a vector. List is inferior in every way unless:
  * You need pointer stability on mutation
  * Your lists will be huge and you expect to modify the middle a lot

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc [hunk 4]
```diff
@@ -134,6 +136,76 @@ TEST_F(SerializerTest, EncodesBoolProtoToBytes) {
...
+    proto.value.integer_value = test;
+    ExpectRoundTrip(model, proto, FieldValue::Type::Long);
+  }
+}
+
+TEST_F(SerializerTest, EncodesIntegersProtoToBytes) {
+  // NB: because we're calculating the bytes via protoc (instead of
+  // libprotobuf) we have to look at values from numeric_limits<T> ahead of
+  // time. So these test cases make the following assumptions about
+  // numeric_limits: (linking libprotobuf is starting to sound like a better
```
Comment (wilhuff) [inline Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: FWIW I agree but don't want to necessarily derail you.

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc (hunk 3 above)
Comment (rsgowman) [reply Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: Done. (Not sure why I used a list here... especially given the use of vector in the method immediately above.)

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc (hunk 4 above)
Comment (rsgowman) [reply Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: Yup. I think I'll experiment with this shortly. (Though I may wait for a few more PRs that pile on the tech dept until it becomes impossible to ignore any further.)

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 1 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: If you merge that PR down here I think this one is ready to go.

**Pull request merged**